### PR TITLE
Feature/mcuboot modem dfu

### PR DESCRIPTION
**Pull request by chunfantasy**
This PR is to support modem dfu via mcuboot.

### CODE REVIEW

File: lib/actions/mcubootTargetActions.js [hunk 1]
```diff
@@ -115,13 +117,33 @@ export const pickSerialPort = serialports => {
...
+export const pickSerialPort2 = serialports => {
+    const platform = process.platform.slice(0, 3);
+
+    switch (platform) {
+        case 'win':
+            return serialports.find(s => (/MI_0[23]/.test(s.pnpId)));
+        case 'lin':
+            return serialports.find(s => (/-if0[23]$/.test(s.pnpId)));
+        case 'dar':
+            return serialports.find(s => (/[23]$/.test(s.comName)));
```
Comment (bencefr) [inline lib/actions/mcubootTargetActions.js]: `/[34]$/`

File: lib/actions/mcubootTargetActions.js [hunk 2]
```diff
@@ -115,13 +117,33 @@ export const pickSerialPort = serialports => {
...
+        default:
+    }
+
+    return undefined;
```
Comment (bencefr) [inline lib/actions/mcubootTargetActions.js]: `return {}` so .comName doens't fail later.

File: lib/components/ModemUpdateDialogView.jsx [hunk 3]
```diff
@@ -70,16 +71,25 @@ class ModemUpdateDialogView extends React.Component {
...
+            progressPercentage,
+            progressStep,
             onCancel,
+            isMcuboot,
         } = this.props;
         if (this.intervalId && (isWritingSucceed || isWritingFail)) {
             clearInterval(this.intervalId);
         }
         const { timer } = this.state;
+        let preogressStepStatus = '';
```
Comment (bihanssen) [inline lib/components/ModemUpdateDialogView.jsx]: s/preogress/progress/

File: lib/components/ModemUpdateDialogView.jsx [hunk 4]
```diff
@@ -90,16 +100,38 @@ class ModemUpdateDialogView extends React.Component {
...
+                                now={progressPercentage}
+                                label={`${progressPercentage}%`}
+                            />
+                        )}
                     </Form.Group>
                     <Form.Group>
-                        {isWritingSucceed && (
+                        {isMcuboot && !isWriting && !isWritingSucceed && !isWritingFail && (
+                            <Alert variant="warning">
+                                <p>You are now proceeding modem DFU via MCUboot.</p>
```
Comment (bihanssen) [inline lib/components/ModemUpdateDialogView.jsx]: s/proceeding/performing/

File: lib/components/ModemUpdateDialogView.jsx [hunk 5]
```diff
@@ -90,16 +100,38 @@ class ModemUpdateDialogView extends React.Component {
...
+                                <p>The device will be recovered if you proceed to write.</p>
```
Comment (bihanssen) [inline lib/components/ModemUpdateDialogView.jsx]: I suggest s/recovered/overwritten/

File: lib/actions/mcubootTargetActions.js (hunk 1 above)
Comment (chunfantasy) [reply lib/actions/mcubootTargetActions.js]: Updated

File: lib/actions/mcubootTargetActions.js (hunk 2 above)
Comment (chunfantasy) [reply lib/actions/mcubootTargetActions.js]: Updated

File: lib/components/ModemUpdateDialogView.jsx (hunk 3 above)
Comment (chunfantasy) [reply lib/components/ModemUpdateDialogView.jsx]: Updated

File: lib/components/ModemUpdateDialogView.jsx (hunk 4 above)
Comment (chunfantasy) [reply lib/components/ModemUpdateDialogView.jsx]: Updated

File: lib/components/ModemUpdateDialogView.jsx (hunk 5 above)
Comment (chunfantasy) [reply lib/components/ModemUpdateDialogView.jsx]: Updated

File: lib/actions/modemTargetActions.js [hunk 6]
```diff
@@ -134,3 +136,120 @@ export const performUpdate = () => (dispatch, getState) => {
...
+
+export const performUpdate = () => (dispatch, getState) => {
+    dispatch(writingStartAction());
+    dispatch(processUpdateAction(MODEM_DFU_STARTING));
+    logger.info('Modem DFU starts to write...');
+
+    const { modemFwName: fileName } = getState().app.modem;
+    const serialNumber = Number(getState().app.target.serialNumber);
+    logger.info(`Writing to device with serial number: ${serialNumber}`);
+    logger.info(`Writing to device with file: ${fileName}`);
```
Comment (bencefr) [inline lib/actions/modemTargetActions.js]: I suggest to have simplify the two lines by fixing the grammar:
`Writing ${fileName} to device ${serialNumber}`

**Pull request merged**